### PR TITLE
Use project name as setting prefix (fixes #448)

### LINKED ISSUE #448: Settings names based on project name
One of the drawbacks of using cliquet in your app is that your settings names are prefixed with ``cliquet.`` (see [previous discussion](https://github.com/mozilla-services/cliquet/issues/318)).

Same for the management command, called ``cliquet`` to migrate ([see related kinto issue](https://github.com/Kinto/kinto/issues/129))

What if, in Cliquet, we would read the settings with a *dynamic* namespace based on the project name (this only one would be ``cliquet.project_name`` though) ?
Instead of having to define ``cliquet.http_host`` in the application, we would define ``readinglist.http_post``...

Didn't look about possible implementation so far. Just thinking loudly.

### PR DESCRIPTION
**Pull request by leplatrem**
As suggested by @ametaireau, this re-implementation of #470 is ~~much more sweet and light~~ not after discussing :)

Thoughts ? /cc @Natim

### CODE REVIEW

File: cliquet/initialization.py [hunk 1]
```diff
@@ -367,6 +368,65 @@ def on_new_response(event):
...
+                is_already_set = 'basicauth' in settings['multiauth.policies']
+                if asbool(settings.pop(old)) and not is_already_set:
+                    settings['multiauth.policies'] += ' basicauth'
+            else:
+                settings[new] = settings.pop(old)
+
+    config.add_settings(settings)
+
+
+def handle_project_name_prefix(settings):
```
Comment (ametaireau) [inline cliquet/initialization.py]: a description here would be very welcome :)

File: cliquet/initialization.py [hunk 2]
```diff
@@ -367,6 +368,65 @@ def on_new_response(event):
...
+    for key, value in settings.items():
+        if key.startswith('cliquet.'):
+            prefixed_key = re.sub(r"^cliquet", project_name, key)
+            settings.setdefault(prefixed_key, value)
+
+    # Then, use <project_name>.* for unprefixed settings.
+    for key, value in settings.items():
+        if key.startswith(project_name + '.'):
+            unprefixed_key = re.sub(r"^.+\.", "", key)
+            settings[unprefixed_key] = value
```
Comment (ametaireau) [inline cliquet/initialization.py]: Is there a way to not loop on the settings three times?

Comment (leplatrem) [inline cliquet/initialization.py]: Maybe. But I could not find it.

**Comment by leplatrem** [comment]
After discussing : 

* [ ] Remove prefix from default settings
* [ ] Remove prefix when reading settings at init
* [ ] Remove prefix when reading in app
* [ ] Warn if both `cliquet.*` and `readinglist.*` are defined

File: cliquet/initialization.py [hunk 3]
```diff
@@ -367,6 +368,43 @@ def on_new_response(event):
...
+
+    project_name = settings['project_name']
+
+    def _prefixed_keys(key):
+        unprefixed = key.split('.', 1)[1] if '.' in key else key
+        project_prefix = project_name + '.' + unprefixed
+        cliquet_prefix = 'cliquet.' + unprefixed
+        return unprefixed, project_prefix, cliquet_prefix
+
+    # # Fill settings with default values if not defined.
```
Comment (Natim) [inline cliquet/initialization.py]: nit: Remove one #

File: cliquet/storage/memory.py [hunk 4]
```diff
@@ -154,7 +154,7 @@ class Memory(MemoryBasedStorage):
 
     Enable in configuration::
 
-        cliquet.storage_backend = cliquet.storage.memory
+        myapp.storage_backend = cliquet.storage.memory
```
Comment (Natim) [inline cliquet/storage/memory.py]: project_name instead of myapp?

**Comment by leplatrem** [comment]
> Note: See Pyramid source. Not very convincing : https://github.com/Pylons/pyramid/blob/master/pyramid/config/settings.py#L64-L149

File: cliquet/storage/memory.py (hunk 4 above)
Comment (leplatrem) [inline cliquet/storage/memory.py]: or maybe leave cliquet like in other backends :)

**Comment by ametaireau** [comment]
Looks good to me. r+

**Pull request merged**